### PR TITLE
[AIRFLOW-1644] Parse comments in hive dry_run

### PR DESCRIPTION
**Pull request by cristianfr**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1644

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

HiveOperator dry_run recognizes as a set up any hql preceeded by
comments. As such by adding a comment one can execute arbitrary code
during a dry_run. Modifying hive_hook test_hql to properly identify
statements and ignore comments.

### Tests
- [x] My PR adds the following unit tests: HivePrestoTest:test_hive_dryrun_with_comments

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: airflow/hooks/hive_hooks.py [hunk 1]
```diff
@@ -227,44 +227,53 @@ def test_hql(self, hql):
...
-                        end = min(l+3, len(query.split('\n')))
-                        context = '\n'.join(query.split('\n')[begin:end])
-                        self.log.info("Context :\n %s", context)
+                query_set.append(('insert', query_original, comments))
+            elif query.startswith('drop table'):
+                query_set.append(('drop', query_original, comments))
+            else:
+                query_set.append(('other', query_original, comments))
+
+        for (mode, query, comments) in query_set:
```
Comment (aoen) [inline airflow/hooks/hive_hooks.py:257]: Let's make queries a NamedTuple

**Comment by aoen** [comment]
This is a pretty high risk change since this code is pretty hacky and not well tested, and the diff is decently sized. The proposed changes are pretty hacky to (largely because they extend some of the bad patterns of the existing code). I know it sucks to hear, but I'm not sure the value added is high enough here until we have more testing. Let me know if you disagree.

**Comment by cristianfr** [comment]
There's currently a bug that allows to run arbitrary hql in a dry_run as exposed in the test. (edit: code -> hql)

**Comment by cristianfr** [comment]
What's the actionable suggestion?
- Remove dry_run
- Add a sql parser to airflow (the reason it's hacky it's because we lack a sql parser which would make a much bigger diff)
- Add more tests that should have been there but were not there?

**Comment by aoen** [comment]
I get that, but I think the potential damage caused by some breakage in one of the most commonly used hooks isn't worth fixing this bug until we have better testing around it (and ideally refactor the code too). Maybe we can make a new function that just prints the hql instead of dry running and that can be used in the meantime as an alternative (and that can e.g. be passed into a gateway hive cli if the user wants to dry run).

**Comment by aoen** [comment]
Since DAGs can depend on dry_run not sure that we can just flat out remove it, adding a warning would probably be the best idea for now. The proper fix would be using a proper hive parser as you suggested, as well as adding thorough tests (e.g. testing edge cases around line length), and actually validating the results of the existing tests instead of just running them and checking if they had errors. Definitely a lot of work.

**Comment by cristianfr** [comment]
Sounds reasonable. I'll close then seems it seems there should be another ticket addressing this more directly as you mention.

**Pull request closed without merging**